8345314: Add a red–black tree as a utility data structure

caspernorrbin opened the pull request:
Hi everyone,

This effort began as an exploration of replacing the current NMT treap with a red-black tree. Along the way, I discovered that others were also interested in having a general-purpose tree structure available within HotSpot.

The red-black tree is designed to serve as a drop-in replacement for the existing NMT treap, keeping a nearly identical interface. However, I’ve also added a few additional requested features, such as an iterator.

Testing builds off the treap tests, adding a few extra that inserts/removes and checks that the tree is correct. Testing uses the function `verify_self`, which iterates over the tree and checks that all red-black tree properties hold. Additionally, the tree has been tested in vmatree instead of the treap without any errors.

For those who may want to revisit the fundamentals of red-black trees, [Wikipedia](https://en.wikipedia.org/wiki/Red%E2%80%93black_tree) offers a great summary with tables covering the various balancing cases. Alternatively, your favorite data structure book could provide even more insight.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8345314](https://bugs.openjdk.org/browse/JDK-8345314): Add a red–black tree as a utility data structure (**Enhancement** - P4)


### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) Review applies to [b2835590](https://git.openjdk.org/jdk/pull/22360/files/b2835590db6bf3bd8dbf65eb282b9971cc03307f)
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) Review applies to [b7219a93](https://git.openjdk.org/jdk/pull/22360/files/b7219a93ab644f4c8be30d4220b5a4d0c7b92d57)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22360/head:pull/22360` \
`$ git checkout pull/22360`

Update a local copy of the PR: \
`$ git checkout pull/22360` \
`$ git pull https://git.openjdk.org/jdk.git pull/22360/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22360`

View PR using the GUI difftool: \
`$ git pr show -t 22360`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22360.diff">https://git.openjdk.org/jdk/pull/22360.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22360#issuecomment-2511893784)
</details>

jdksjolen commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,1041 @@
+    RBNode* _parent;
+    RBNode* _left;
+    RBNode* _right;
+    Color _color;
```
Actually, we could even use the MSB of `_black_height` to encode the red/black-ness of a node. 0 bytes used, then.

jdksjolen commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,1041 @@
+    const K& key() const { return _key; }
+    V& val() { return _value; }
+
+    // Unsafe operation, user needs to guarantee that:
+    // All nodes less than the previous key is also less than the new key.
+    // All nodes greater than the previous key is also greater than the new key.
+    void update_key(K& new_key) {
+    #ifdef ASSERT
+      assert(verify_key_update(new_key), "New key does not satisfy current tree conditions");
+    #endif // ASSERT
+      _key = new_key;
+    }
```
If it's needed, save it for a future PR. We'll take the pain of getting it past reviewers then.

jdksjolen reviewed (commented):
Initial review

caspernorrbin commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,735 @@
+    // Black node removed, balancing needed
+    RBNode* parent = node->_parent;
+    while (parent != nullptr) {
+      RBNode* sibling = node->is_left_child() ? parent->_right : parent->_left;
```
Once we reach here, node is a leaf node and black, which means that a sibling has to exist to not cause a black-violation where the path through the sibling has one less black node.
I added an explaining comment + changed the function name to be clearer.

dholmes-ora commented on src/hotspot/share/utilities/rbTree.hpp:
Just a drive-by comment but why is this all in the hpp file instead of being split into hpp (and maybe .inline.hpp) and cpp files?

caspernorrbin replied:
The entire tree is templated, so we can't easily move the implementation to a cpp file

theRealAph replied:
You can, by using template instantiation. That would be an improvement over putting it all in the header file, IMHO.

jdksjolen replied:
I'm a bit surprised that it's contentious to have the whole tree in the header file when we do the same with other templated types such as `GrowableArray`.

caspernorrbin replied:
@theRealAph I don't think that would be possible either. We can't possibly know every type combination to instantiate, especially since we're asking users to provide their own comparator and allocator through the template.

And like Johan said, we already have templated header-only structures, like `GrowableArray` and `ConcurrentHashTable`.

theRealAph replied:
I think the case for making the whole thing in a header depends on how much code there is, that's all.

coleenp replied:
concurrentHashTable uses the trick of inlining a lot of methods in concurrentHashTable.inline.hpp and concurrentHashTableTasks.inline.hpp that are included in the .cpp callers.  This might work for this also.

dholmes-ora replied:
FWIW this was a drive by comment and it just surprised me to see everything in the hpp file. I don't know GrowableArray enough to recognise it has a similar approach.

coleenp replied:
Also one advantage of putting much of the implementation in the .inline.hpp file is that it's easy to see how to use this if only a small number of lines of code are in the hpp file.

tstuefe replied:
@theRealAph 

> Yes of course we can. Users add their own instantiations as required. I'm not saying you must do this, but I am saying it's not hard.

But for that the compiler must see the implementation of the functions, right? So, we cannot move the implementation into a cpp file?

We can move it into an inline header, like we usually do. Which serves much of the same purpose.

tstuefe replied:
+1 for hpp/.inline.hpp scheme.

theRealAph replied:
> @theRealAph
> 
> > Yes of course we can. Users add their own instantiations as required. I'm not saying you must do this, but I am saying it's not hard.
> 
> But for that the compiler must see the implementation of the functions, right? So, we cannot move the implementation into a cpp file?

Sure we can. That's what explicit template instantiation does. The cpp file contains the definition of the functions plus the explicit instantiations of whatever template types are needed.

As an example, imagine that the rbtree.inline.hpp implementation were moved to rbtree.cpp. Then there could be a number of  explicit template instantiations in rbtree.cpp like this one:

`template class RBTree<concreteK, concreteV, concreteCOMPARATOR, concreteALLOCATOR>;`

This generates all of the code for the template functions. But it's generated once and called, rather than inlined. The only stuff in the header would be the declarations.

NB: Again, i am _not_ saying that we should do this, but I am saying that it's an option for large template classes.

tstuefe replied:
@theRealAph Just to be sure I understand, this means that if I want to add a new variant at a later point, I need to add its explicit instantiation to this cpp file, right? Including giving it all the headers it needs to resolve the template types.

I guess this is what others meant by "we don't know the types others use upfront". We could do this, its just a bit of a hassle. OTOH, it could reduce template bloat.

theRealAph replied:
> @theRealAph Just to be sure I understand, this means that if I want to add a new variant at a later point, I need to add its explicit instantiation to this cpp file, right? Including adding all the headers it needs to resolve the new template types.

Yes, indeed that's right.

> I guess this is what others meant by "we don't know the types others use upfront". We could do this, its just a bit of a hassle. OTOH, it could reduce template bloat.

No big deal at all, I would have thought. A practical downside might be that a bunch of unused convenience methods would be generated, but a linker can delete many of those,

offamitkumar commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,724 @@
+    }
+
+  private:
+    void color_black() {
+      _color = BLACK;
+    }
+
+    void color_red() {
+      _color = RED;
+    }
```
```suggestion
    void set_color_black() {
      _color = BLACK;
    }

    void set_color_red() {
      _color = RED;
    }
```

afshin-zafari commented on test/hotspot/gtest/utilities/test_rbtree.cpp:
```diff
@@ -0,0 +1,435 @@
+    // Create random values
+    for (int i = 0; i < size; i++) {
+      int r = os::random() % size;
+      allocations.append(r % size);
```
`... % size` is already done in previous line.

afshin-zafari reviewed (commented):
Thank you for the nice implementation. The code is so neat and clean.
One point I want to bring here is testing the RBTree if it is really balanced or not. This is the main characteristic of the RBTree. It can be done, for example, by a new API as `RBTree.get_max_depth()` which is expected to be $\approx log_2(n)$ if the tree has $n$ nodes.

caspernorrbin commented:
Thank you everyone for the comments!

I split the tree into `rbTree.hpp` and `rbTree.inline.hpp`, which is similar to how `concurrentHashTable` works. The original file is now much more readable, with some additional added comments to the public functions. All the internal functions is hidden away in the inline file.

I also added a check to `verify_self()` which ensures that the tree is balanced, i.e. the upper bound of $2log_2(n+1)$ is never exceeded, which should have been checked for previously. Thanks @afshin-zafari for pointing it out.

jdksjolen commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,510 @@
+#include "utilities/powerOfTwo.hpp"
+#include "utilities/rbTree.hpp"
+
+// ----------------
+// RBNode functions
+// ----------------
```
No need for fancy comment

caspernorrbin replied:
Removed them :)

jdksjolen commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,510 @@
+}
+#endif // ASSERT
+
+// ------------------
+// Iterator functions
+// ------------------
```
No need for fancy comment

jdksjolen commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,310 @@
+
+private:
+  template<bool Forward>
+  class IteratorImpl : public StackObj {
```
Make it `NONCOPYABLE` as a safety measure.

jdksjolen commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,496 @@
+    return false;
+  }
+
+  if (node->_left != nullptr && node->_right != nullptr) { // node has two children
+    // Copy the k/v from the in-order successor and delete that node instead
+    RBNode* curr = node->_right;
+    while (curr->_left != nullptr) {
+      curr = curr->_left;
+    }
+    node->_key = curr->key();
+    node->_value = curr->val();
+
+    node = curr;
+  }
```
@tstuefe,

Seems like nodes are not stable. To quote Wikipedia:

>- When the deleted node has 2 children (non-NIL), then we can swap its value with its in-order successor (the leftmost child of the right subtree), and then delete the successor instead. Since the successor is leftmost, it can only have a right child (non-NIL) or no child at all. 

This is quite unfortunate, as it becomes very difficult (impossible?) to have intrusive RB-trees with this. 

@caspernorrbin,

Finding information on this in the literature seems like it takes quite a bit of digging. Can't we replace this with swapping everything *but* the key and value?

```c++
  if (node->_left != nullptr && node->_right != nullptr) { // node has two children
    RBNode* curr = node->_right;
    while (curr->_left != nullptr) {
      curr = curr->_left;
    }
  // Swap parent, left, right, and color
  std::swap(curr->_parent, node->_parent);
  std::swap(curr->_left, node->_left);
  std::swap(curr->_right, node->_right);
  std::swap(curr->_color, node->_color);
  // Swap the parent's child pointer
  if (curr->_parent->_left == node) curr->_parent->_left = curr;
  if (curr->_parent->_right == node) curr->_parent->_right = curr;
  // Set the children's parent pointers 
  curr->_left->_parent = curr; curr->_right->_parent = curr;

  // and then the identical writes for node
  if (node->_parent->_left == curr) node->_parent->_left = node;
  if (node->_parent->_right == curr) node->_parent->_right = node;
  // Set the children's parent pointers 
  node->_left->_parent = node; node->_right->_parent = node;

  // Done, they've swapped places in the tree.
  node = curr;
  }
```

caspernorrbin replied:
Nodes are now stable! This was not something I had considered, but it's possible with what Johan suggested above. Previously, keys and values were swapped around between nodes during deletion but now we instead re-point the node's pointers instead. With this the key and value stays constant in a node through its lifetime, given the user does not themself change the value.

tstuefe replied:
@caspernorrbin @jdksjolen Excellent, thank you. 

For the Doubting Thomas I am, could you write a little gtest like this:

- node (key = size) value don't matter I guess.
- malloc n times, with random sizes, so variable sized, but large enough to hold at least a node structure
- then in a loop, add or remove a randomly chosen malloc block to the tree.

If we survive that, I am happy and will use this in the future in Metaspace. Probably also in other areas.

tstuefe replied:
I realize that a simpler test would be to just create random nodes (key=size, value=this node's address), then add/remove randomly while checking that the node address still equals the node value.

caspernorrbin replied:
(Forgot to mention) I added a gtest that allocates and stores nodes together with their keys, then performs a bunch of deletions to then check that the value of the nodes remain unchanged. Can add another with addresses too to be extra safe

jdksjolen commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,313 @@
+
+};
+
+template <MemTag mt>
+class RBTreeCHeapAllocator {
+public:
+  void* allocate(size_t sz) {
+    void* allocation = os::malloc(sz, mt);
+    if (allocation == nullptr) {
+      vm_exit_out_of_memory(sz, OOM_MALLOC_ERROR,
+                            "red-black tree failed allocation");
+    }
+    return allocation;
+  }
+
+  void free(void* ptr) { os::free(ptr); }
+};
```
Some more thoughts: It's not the allocator that should run the destructor, but the RBTree itself. It should probably enforce that the key is a trivial type also.

caspernorrbin replied:
We now assert that the key is trivially destructible and the value's destructor is called when removing a node

caspernorrbin commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,496 @@
+  if (_right != nullptr) {
+    _right->visit_in_order_inner(f);
+  }
+}
```
I thought the recursive versions looked neater, but changed them now to the iterative version with pre-sized GAs

jdksjolen replied:
I agree, they do look neater. The pre-sized GAs are also kinda costly, often you want to keep your allocated GA around for future calls (found this out the hard way), whilst stack allocations are just gonna be kept around and it's unlikely that we'll hit the stack ceiling but... Let's just be safe rather than sorry.

caspernorrbin commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,496 @@
+    } else {             // k > key
+      curr = curr->_right;
+    }
+  }
```
`find_node` either returns the node matching the key or `nullptr`, where here we go down the tree and stay on the parent if the exact key wasn't found in order to insert below it. This would be lost if using `find_node`

jdksjolen replied:
Good point, thanks!

caspernorrbin commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,496 @@
+      return;
+    }
+
+    RBNode* uncle = parent->is_left_child() ? grandparent->_right : grandparent->_left;
+    if (is_black(uncle)) { // Parent is red, uncle is black
```
A NIL node (nullptr) is considered black, so `is_black(nullptr)` is valid and returns true.

jdksjolen replied:
Oh right, could've found that out if I read the definition of `is_black`, thanks!

jdksjolen commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -29,13 +29,16 @@
 //     - an int > 0 when a > b
 // ALLOCATOR must check for oom and exit, as RBTree currently does not handle the
 // allocation failing.
+// Key needs to be of a type that is trivially destructable
```
destruct**i**ble

jdksjolen reviewed (commented):
Could you add a comment saying that nodes are address stable?

jdksjolen reviewed (approved):
One unclear comment, but I am happy with this now.

Thank you!!!

caspernorrbin commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,313 @@
+// allocation failing.
+// Key needs to be of a type that is trivially destructible.
+// The tree will call a value's destructor when its node is removed.
+// Nodes are address stable and will not change during its lifetime unless modified by the user.
```
Removed the last part. I was trying to say that the content of a node never changes (unlike previously) unless the user explicitly changes the value.

jdksjolen replied:
Yeah, I think that can be more confusing than not mentioning it :).

tstuefe commented:
I am playing around with this tree to replace the backing storage of the compilation memory statistics. Some notes:

Why does the tree own the allocator? Why is its lifetime bound to the tree? This restricts the usefulness of an allocator. Two trees have two different allocator instances, but since we have no way of telling apart the caller tree instances in the allocator code, there is no point in that. We might just as well make the allocator AllStatic then and call via `ALLOCATOR::allocate`.

Consider this scenario: I want to have two short-lived trees, and I want the nodes to reside in physically different pools. I want to use the same allocator and tree classes, though, so template parameters must be the same.

There are two ways to do this. The most obvious would be to keep a reference to an allocator instance that the caller hands in via constructor. So, make _allocator a pointer, and if omitted by the caller use your own C-heap allocator. Otherwise, use the caller's allocator.

Another way would be to make the Allocator AllStatic and give each tree a caller-defined token (e.g. void*, think pthread_create), then hand this into all allocator functions. Tbh, the first variant sounds easier.

The node tree must also handle OOM because maybe I don't want to exit on OOM. An OOM from the node allocator itself may be a normal and expected behavior. Eg, consider this: I want to use nodes from a limited pre-allocated buffer, and I want to handle buffer OOM by, e.g., deleting the leftmost nodes and retrying. Example, keeping the N largest "somethings" during a VM run within a limited buffer. In that case, an allocator OOM would be the most natural and understandable way to implement this. As the tree works now, we need to predict and forestall an OOM by counting nodes *before* adding a new one. That is awkward and maybe not even possible if the allocator OOM is not easily predictable (think variable-sized nodes, or allocator fragmentation issues).

tstuefe commented:
Here is another idea, but that may be a stretch goal. 

I would love it if we could create a fast copy of a tree (think NMT snapshotting) by just copying the node memory. Instead of having to laboriously walk the tree structure and rebuild the whole tree.

Obviously, this only works with an allocator that pools the nodes from a confined pool. But okay, I can provide such an allocator. The second problem is that the tree nodes are wired up with absolute pointers. So, copying the node memory does not buy much, since the root node copy still points to nodes in the original pool. 

So, my proposal would be to let the allocator define an allocation base address. Like this:

```
class MyCopyableAllocator {
  Node my_pool[100000];
  ...
  Node* get_node_base_address() const override { return my_pool; }
  };
```
And in the tree, when building a new node, subtract the base address from the references. When walking, add the base address to find the next node. 

For a simple C-heap allocator, the `get_node_base_address` function can just return NULL.

tstuefe commented:
Thinking about this some more, the fact that tree nodes are allocated internally and are separate entities from the key/value holders precludes the using this tree for memory management (address stable or not). It works, but the separate allocation defeats the purpose. It would be good if, on `upsert`, I could hand in optionally the future node memory itself, and the node should then be placed into/at the start of this memory.

tstuefe commented:
> Thinking about this some more, the fact that tree nodes are allocated internally and are separate entities from the key/value holders precludes the using this tree for memory management (address stable or not). It works, but the separate allocation defeats the purpose. It would be good if, on `upsert`, I could hand in optionally the future node memory itself, and the node should then be placed into/at the start of this memory. And it must be able to deal with the handed-in node memory be variable-sized, of course.

Or, much simpler, let me construct the node itself and hand it into the tree.

jdksjolen commented:
Hi @tstuefe ,

>Consider this scenario: I want to have two short-lived trees, nodes live longer than their trees they temporarily occupy, and I want the nodes to reside in physically different preallocated pools. I want to use the same allocator and tree classes, though, so template parameters must be the same.

>There are two ways to do this. The most obvious would be to keep a reference to an allocator instance that the caller hands in via constructor. So, make _allocator a pointer, and if omitted by the caller use your own C-heap allocator. Otherwise, use the caller's allocator.

Or you have a reference as the template type and you alter the constructor of the `RBTree` to take such a type to construct its own allocator with.

It seems like you want an intrusive RBTree. That's really no problem, it just means that we have to extend the API slightly and make it so that you can avoid using an allocator.

I do want an allocator instance in the tree, because then you can make very nice things such as a `IndexWithFreeList` allocator that doesn't need protection from concurrent writes. You could also, for example, avoid the cost of freeing separate elements and eliminate the whole tree at once when the lifetime of the tree has ended (since the lifetime of the allocator is identical). In short: When your tree isn't intrusive, it's very nice to have a custom allocator.

For your use case, you basically need a very similar addition to the API. It'd look something like this in pseudo-code:

```c++
// I hold the data, but the node holds the key
struct IntrusiveHolder {
  int data;
  RBNode node;
  static IntrusiveHolder* cast_to_self(RBNode* node) ; // offset_of() magic
};

struct Compare {
  int cmp(int a, int b);
}

template<typename K, typename V, typename Cmp, typename Allocator>
struct RBTree{
public:
  enum Dir { left, right };
  // Describes the parent node and whether the next node is to be inserted on the left or the right.
  struct InsertionCursor {
   RBNode* parent; Dir dir;
   RBNode* position(); // Get the node (potentially nullptr)
  };

  // Walks the tree in the obvious order, using Comparator::cmp overload to find the place to insert.
  // However does not allocate.
  InsertionCursor find_insertion_point(K k);
  // Inserts the node at the position and rebalances the tree.
  void insert_at_cursor(InsertionCursor, RBNode* node);

  // For removals we need to unlink the node from the tree, but keep it alive and then just return the pointer.
  RBNode* remove(K k);
};

// An intrusive RBTree doesn't store any V, and it should not use a custom allocator.
struct Empty{};
struct NoopAlloc {
  void* allocate(size_t sz) {
    assert(false, "no");
    return nullptr;
  }
  void free(void* ptr) {
    assert(false, "no");
    return;
  }
};
template<typename K, typename Cmp>
using IntrusiveRBTree = RBTree<K, Empty, Cmp, NoopAlloc>;

void insert_and_then_remove(IntrusiveRBTree<int, Compare>& tree, int key, int data) {
  InsertionCursor cursor = tree.find_insertion_point(key);
  if (cursor.position() != nullptr) {
    IntrusiveHolder::cast_to_self(cursor.position())->data = data;
  } else {
    // Our custom allocation scheme is just malloc here
    IntrusiveHolder* place = os::malloc(sizeof(IntrusiveHolder), mtTest);
    new (place) IntrusiveHolder(data, RBNode());
    tree.insert_at_cursor(cursor, key, place->node);
  }
  
  RBNode* node = tree.remove(key);
  if (node != nullptr) {
    os::free(IntrusiveHolder::cast_to_self(node)); // Find the outer structure which malloc knows how to free.
  }
}
```

We could also change the comparator so that it takes `(RBNode*, RBNode*)`, then the key can also be stored in the `IntrusiveHolder` and we can make `K = Empty`. The details on that are slightly fuzzy.

Casper already is working on supporting an intrusive RBTree with an API similar to what I sketched out above, but I'd like it if we can integrate the current version of the tree as it is. With this change we can already remove the NMT Treap and replace it, so that's a nice use case. Plus, there'll always be new features, we can't wait with integration until all possible features are implemented :).

tstuefe commented:
Hi @jdksjolen, @caspernorrbin 

sorry for the delay, work is piled up.

> 
> It seems like you want an intrusive RBTree. That's really no problem, it just means that we have to extend the API slightly and make it so that you can avoid using an allocator.

intrusive - I did not know this term, even though I wrote containers like this all my life, go figure :-). 

Yes, that's precisely what I want.

- I want to create/store nodes myself 
- I want to avoid copying node content; in fact, I don't want to even bother with assignment operators.
- I want to be able to add child nodes to a node to handle duplicates if my tree allows duplicates (which is of no concern to the tree, but since these child nodes should live in the same pool, again, my wish for wanting to create nodes myself)

All of this points to a tree where I manage the node myself. So, have a function to find a reference to an existing node by key and have a function to hand in a node for storage. Pretty much like you line it out in your example:

```
MyTree::add_my_bulky_and_possible_varsized_node(Node* n) {
	- find node n2 with key n->key
		- found? Add n as child node under n2
		- not found? add n as new node
}
```

I have two immediate examples where I want to use this RB tree: the reworked c2 memory statistic (in fact, I withhold the patch to wait for this RB tree to drop) and - somewhat later - the metaspace binary tree dictionary replacement. Both cases would be perfectly served with an intrusive key that allows me to handle duplicates.


> 
> I do want an allocator instance in the tree, because then you can make very nice things such as a `IndexWithFreeList` allocator that doesn't need protection from concurrent writes.

You lost me here

> You could also, for example, avoid the cost of freeing separate elements and eliminate the whole tree at once when the lifetime of the tree has ended (since the lifetime of the allocator is identical).

This is not a strong argument though, since when I create the allocator outside the tree, I can let it die in the same fashion when the tree dies. This is what I plan to do. But as you wrote, we should allow both intrusive and non-intrusive.

> Casper already is working on supporting an intrusive RBTree with an API similar to what I sketched out above, but I'd like it if we can integrate the current version of the tree as it is. With this change we can already remove the NMT Treap and replace it, so that's a nice use case. Plus, there'll always be new features, we can't wait with integration until all possible features are implemented :).

Hmm, okay. I'll give the patch a closer look later today and then approve it unless I find something.

xmas92 reviewed (commented):
Good work on the RB-tree implementation.

I am no expert in RB-trees but the implementation does not look obviously incorrect. And the gtest and verification looks comprehensive. (Only thing I might have extended is to randomness to also check the visitor and iterators. Something like generating a random ordered set of keys in a different container, insert them in some random order, maybe including insertions and removals of other keys and then check that the visitor and iterator visit this set correctly.) 

The RB-tree logic looks good to me.

Except for some small nits most of my feedback is about the API and the iteration.

I would like to avoid GrowableArrays in the implementation, but maybe I am missing a reason why they are preferable. If we could design the iterators to work more like standard C++ iterators (at least simplified but compatible versions) would make them more useful and ergonomic. There is also an opportunity to use such iterators as cursors and create a nice interface for using this tree in many different scenarios.

The second is if we should add more const version of the APIs to make it easier to work with the tree in const context (e.g. pass around the tree as a `const& Tree`). At least `const V* find(const K& key) const` seems useful to have a const `contains`.

Overall I think this is nice. But I would like some feedback on my comments w.r.t. the API before approving.

jdksjolen commented:
> Hi @jdksjolen, @caspernorrbin
> 
> sorry for the delay, work is piled up.
> 
> > It seems like you want an intrusive RBTree. That's really no problem, it just means that we have to extend the API slightly and make it so that you can avoid using an allocator.
> 
> intrusive - I did not know this term, even though I wrote containers like this all my life, go figure :-).
> 
> Yes, that's precisely what I want.
> 
>     * I want to create/store nodes myself
> 
>     * I want to avoid copying node content; in fact, I don't want to even bother with assignment operators.
> 
>     * I want to be able to add child nodes to a node to handle duplicates if my tree allows duplicates (which is of no concern to the tree, but since these child nodes should live in the same pool, again, my wish for wanting to create nodes myself)
> 
> 
> All of this points to a tree where I manage the node myself. So, have a function to find a reference to an existing node by key and have a function to hand in a node for storage. Pretty much like you line it out in your example:
> 
> ```
> MyTree::add_my_bulky_and_possible_varsized_node(Node* n) {
> 	- find node n2 with key n->key
> 		- found? Add n as child node under n2
> 		- not found? add n as new node
> }
> ```
> 
> I have two immediate examples where I want to use this RB tree: the reworked c2 memory statistic (in fact, I withhold the patch to wait for this RB tree to drop) and - somewhat later - the metaspace binary tree dictionary replacement. Both cases would be perfectly served with an intrusive key that allows me to handle duplicates.

This requires the tree to have nodes `<` strictly on the left, and `>=` strictly on the right (as opposed to `>` strictly on the right), or the opposite. I have no clue what having duplicates does to a RBTree when rebalancing. If it's that important to have duplicates, then why have them live inside the tree anyway? Just link them together with an intrusive linked list.

```c++
struct Foo {
  RBNode rb; // with some K as key
  Foo* next;
  struct { /* ... */ } data;
};

void a_m_b_a_p_v_n(Foo* n) {
  Cursor c = tree.find_insertion_point(n);
  if (c.exists()) {
    Foo* f = Foo::cast_to_self(c.node());
    n->next = f->next;
    f->next = n->next;
  }
  tree.insert(c, n);
};
```

Just do that. Let's us entirely skip the headache of the rest of it, at the cost of 8 bytes.
 
> > I do want an allocator instance in the tree, because then you can make very nice things such as a `IndexWithFreeList` allocator that doesn't need protection from concurrent writes.
> 
> You lost me here
> 

A static allocator either needs access to local state via a pointer on allocation, or it needs to have some form of protection against concurrent writes. This code below requires a mutex in `alloc`, because it is static.

```c++
static char my_alloc_area[4096];
static size_t ptr_into_my_alloc_area;
void alloc(size_t sz) {
  // Uses my_alloc_area
}

void bar() {
  Thread t1 = run_thread([]() {
    RBTree<K, V, alloc> tree;
    // Do stuff to tree, causing alloc to be called
  });

  Thread t2 = run_thread([]() {
    RBTree<K, V, alloc> tree;
    // Do stuff to tree, causing alloc to be called
  });
}
```

This code doesn't need a mutex:

```c++
struct Alloc {
  char my_alloc_area[4096];
  // And so on
};

void bar() {
  Thread t1 = run_thread([]() {
    RBTree<K, V, Alloc> tree; // RBTree holds local instance of Alloc
    // Do stuff to tree, causing alloc to be called
  });

  Thread t2 = run_thread([]() {
    RBTree<K, V, Alloc> tree; // And so does this one
    // Do stuff to tree, causing alloc to be called
  });
}
```

> > You could also, for example, avoid the cost of freeing separate elements and eliminate the whole tree at once when the lifetime of the tree has ended (since the lifetime of the allocator is identical).
> 
> This is not a strong argument though, since when I create the allocator outside the tree, I can let it die in the same fashion when the tree dies. This is what I plan to do. But as you wrote, we should allow both intrusive and non-intrusive.
> 

Sure, you can, but now you just provide it the allocator and you don't have to think about it anymore. If you want to extend the lifetime, make the allocator a reference and add a constructor argument parameter pack. It's really QoL both for writing and maintaining the code, IMO.

> > Casper already is working on supporting an intrusive RBTree with an API similar to what I sketched out above, but I'd like it if we can integrate the current version of the tree as it is. With this change we can already remove the NMT Treap and replace it, so that's a nice use case. Plus, there'll always be new features, we can't wait with integration until all possible features are implemented :).
> 
> Hmm, okay. I'll give the patch a closer look later today and then approve it unless I find something.

Thanks!

jdksjolen commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,313 @@
+    return remove(node);
+  }
+
+  // Removes the given node from the tree.
+  // Returns true if the node was successfully removed, false otherwise.
+  bool remove(RBNode* node);
```
Wouldn't that require two traversals down the tree? Would a `enum RemovalResult { Removed, NotFound }` be preferred?

caspernorrbin replied:
`remove(RBNode* node)` doesn't traverse the tree, so it is only ever one traversal. I moved the check to `remove(const K& k)` like you suggested, with an added assumption with a comment and assert that the node has to be valid.

jdksjolen commented on test/hotspot/gtest/utilities/test_rbtree.cpp:
```diff
@@ -0,0 +1,511 @@
+      }
+    };
+
+    constexpr const int up_to = 10;
```
TIL, I usually put both.

jdksjolen commented:
> > > Hi @jdksjolen,
> 
> > > I have two immediate examples where I want to use this RB tree: the reworked c2 memory statistic (in fact, I withhold the patch to wait for this RB tree to drop) and - somewhat later - the metaspace binary tree dictionary replacement. Both cases would be perfectly served with an intrusive key that allows me to handle duplicates.
> > 
> > 
> > This requires the tree to have nodes `<` strictly on the left, and `>=` strictly on the right (as opposed to `>` strictly on the right), or the opposite. I have no clue what having duplicates does to a RBTree when rebalancing. If it's that important to have duplicates, then why have them live inside the tree anyway? Just link them together with an intrusive linked list.
> 
> You misunderstood me - a linked list is exactly what I meant. That is why I talked about a "child node". I did this in metaspace:
> 

Phew! Good to hear that we're talking about the same thing.

> https://github.com/openjdk/jdk/blob/f64f22b360f68df68ebb875bd0ef08ba61702952/src/hotspot/share/memory/metaspace/blockTree.hpp#L89-L96
> 
> I can do that myself, as I wrote, this does not concern the tree. So for the tree we can keep things without duplicates. But for simplicity the easiest solution is for child nodes to be of the Node class itself, because then you can just swap them around and make them the leader node for this key value later. That would require a child Node to be allocated in the same pool as the "real" tree nodes, and have the same type as well. All of that can happen outside the tree. It is, however, an argument for giving the caller the ability to allocate and manage nodes himself and passing them in pre-formed for insertion.
> 
 
Right, and you should get that with the intrusive tree. My pseudo-code that you saw above, unfortunately, does not tell you that.


>Oh, sure. But again, not a strong argument for non-intrusiveness. In my original comment, I argued for having the ability to pass in an allocator instance into the destructor, or alternatively use a token of some sort when calling functions in a AllStatic allocator that identifies the tree.

>But as I wrote, I think we don't have to choose. We can have an allocator instance as optional ctor argument, and let the tree create its own allocator itself if nothing is passed in. Or, you know, just use malloc and free instead.

I think that the major argument for non-intrusive datastructures is that they require less ceremony to use. But as you say, we don't have to choose one or the other, and I don't want us to have to.

tstuefe commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,525 @@
+
+#ifdef ASSERT
+template <typename K, typename V, typename COMPARATOR, typename ALLOCATOR>
+inline bool RBTree<K, V, COMPARATOR, ALLOCATOR>::RBNode::is_correct(
```
Some proposals for better/more useful verification:

1) I would use asserts inside the `Node::is_correct` function. I would not bother returning a bool but assert right away. This is used during tree verification only, and ideally, we want to crash hard and fast right there to have aa point to quickly debug. Arguably, I would rename the function to something like `verify()` (I'll use that name from now on)

2) I would then also not use combined asserts but assert each condition separately.

3) Please use size_t (or uintx, which I prefer but I know others don't) for the number of nodes/number of blacks in the arguments of this function. 

4) I would not pass in limits to be checked by the node. Instead would let Node::verify() increase stats.

5) I would also return longest/shortest path to any leaf, and black-nodes-til-leaf, like this (see below point(6)):

```
void verify(
  size_t& num_nodes,          // total nodes 
  size_t& blacks_until_leaf,  // returns the number of black nodes from here to its leafs
  size_t& longest_leaf_path,  // length of longest to-leaf path from here
  size_t& shortest_leaf_path, // length of shortest to-leaf path from here
  size_t current_depth
  ) const;
```
you also can create a Stat struct and pass a reference to that if you dislike so many arguments.

6) In Node::verify(), I would explicitly check the following red-black tree rules: 
  - If Node is red, it shall have no red childs; (I see you do this already)
  - If a Node has exactly one child, the child must be red (consequently, the node must be black)
  - After calling `Node::verify() `for left and right child recursively, the blacks-until-leaf for right node and the one for the left node should be the same (which is also why we return just one value for `blacks_until_leaf`)
  - the longest leaf path and the shortest leaf path must never be farther apart than factor 2 (so, `shortest_leaf_path <= longest_leaf_path <= shortest_leaf_path * 2 `). That asserts that we are balanced as far as RB trees go.

7) In debug builds, please give the Node class a marker bit: `DEBUG_ONLY(bool _visited)`. Keep inside of the Tree class a `DEBUG_ONLY("bool _expected_visited")` bit. On Node creation/addition, set node->_visited to tree->_expected_visited. Then do this: on each `Tree::verify()` call, flip the tree expected bit, and in `Node::verify()` assert that the current node "_visited" bit is !_expected. Then change node->_visited to _expected. This is to ensure that the tree does not contain duplicates - we should have visited each node exactly once. This will be a thing to look out for when adding caller-allocated nodes as a feature. (the flipping of tree->_expected_visited just lets us avoid resetting the bit in each node after verification is completed)

8) If you do (7), you could also do something like this: write a gtest where all nodes are allocated from an array-like allocator. Now you have an independent way to traverse all nodes (via the array). After tree verification, you can then do one additional sweep through the array and check that every `Node::_visited == Tree::_expected_visited`. This makes sure our iteration process reaches all nodes.

9) outside the tree, once root->verify() returned, just check the num_nodes counter for == Tree::_size. Remove Node::count_nodes(), it's not needed. Now, less code and we have avoided one unnecessary iteration of the tree just to get the node count.

caspernorrbin replied:
Thank you for the many suggestions! I have reworked the verification based on this, including everything in the list.
We assert quicker and separately, values and limits are sent up instead of down to be validated, and I added a lot checks. The visited bit and gtest was a very elegant solution (especially for the future intrusive nodes)

tstuefe commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,313 @@
+public:
+  class RBNode {
+    friend RBTree;
+    friend class RBTreeTest;
```
Is there a reason to be friends of RBTree? Even RBTreeTest?

caspernorrbin replied:
RBTree edits the private fields of the nodes during insertion, and RBTreeTest is for more access to verification in the tests.

tstuefe commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,313 @@
+    V _value;
+
+    enum Color : uint8_t { BLACK, RED };
+    Color _color;
```
Possibly for a future optimization. You could keep the color information as one bit overlayed in one of the pointers. That saves four to eight bytes per Node structure.

The Node pointers have to be aligned to `sizeof(void*)`, so we will always have an alignment shadow of at least two bits. If you encapsulate accessing the tagged pointer correctly, this should not be complicated.

(I believe the kernel RB tree does the same thing)

tstuefe reviewed (commented):
Looked through verification first; will continue tomorrow.

tstuefe commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,525 @@
+  RBNode* old_right = _right;
+
+  _right = old_right->_left;
+  if (old_right->_left != nullptr) {
+    old_right->_left->_parent = this;
+  }
```
```suggestion
  if (_right != nullptr) {
    _right->_parent = this;
  }
```
Alternatively  a `set_parent(node)`, equivalent to the existing `set_child()`, with nullptr handling, would be even better

tstuefe commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,525 @@
+  }
+
+  old_right->_parent = _parent;
+  if (is_left_child()) {
+    _parent->_left = old_right;
+  } else if (is_right_child()) {
+    _parent->_right = old_right;
+  }
```
```suggestion
   _parent->replace_child(this, old_right);
```

tstuefe commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,313 @@
+
+    void replace_child(RBNode* old_child, RBNode* new_child);
+
+    // Move node down to the left, and right child up
```
```suggestion
    // This node down, right child up
    // Returns right child (now parent)
```
(different papers use left/right in different ways, I always get confused. A clear concise comment helps)

tstuefe commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,313 @@
+    // Move node down to the left, and right child up
+    RBNode* rotate_left();
+
+    // Move node down to the right, and left child up
```
```suggestion
    // This node down, left child up
    // Returns left child (now parent)
```

tstuefe commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,525 @@
+  RBNode* old_left = _left;
+
+  _left = old_left->_right;
+  if (old_left->_right != nullptr) {
+    old_left->_right->_parent = this;
+  }
```
```suggestion
  if (_left != nullptr) {
    _left->_parent = this;
  }
```

tstuefe commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,525 @@
+  }
+
+  old_left->_parent = _parent;
+  if (is_left_child()) {
+    _parent->_left = old_left;
+  } else if (is_right_child()) {
+    _parent->_right = old_left;
+  }
```
```suggestion
  _parent->replace_child(this, old_left);
```

tstuefe reviewed (commented):
Checked basic tree rotation functions

tstuefe reviewed (commented):
Some more comments, continue next week

caspernorrbin commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,313 @@
+  public:
+    NONCOPYABLE(IteratorImpl);
+
+    IteratorImpl(const RBTree* tree) : _tree(tree) {
```
After some thoughts, I've opted to remove the iterator code in its entirety. This would be replaced by cursors in the intrusive tree. The cursors have the same functionality and are more versatile, so it makes little sense to keep this in.

caspernorrbin commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -0,0 +1,313 @@
+    RBNode* start = closest_leq(addr);
+    RBNode* end = closest_gt(addr);
+    return Range(start, end);
+  }
```
I believe I got this from the treap, but it is not actively used anywhere. After a second look I feel like this belongs more on the caller's side. If someone would want this in the future, it would be trivial for them to implement instead.

caspernorrbin commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,525 @@
+  }
+}
+
+template <typename K, typename V, typename COMPARATOR, typename ALLOCATOR>
+template <typename F>
+inline void RBTree<K, V, COMPARATOR, ALLOCATOR>::visit_range_in_order(const K& from, const K& to, F f) {
```
I added a `BoundMode` to closest_(leq/gt) to allow for this. Now we use two `closest_gt(INCLUSIVE)` calls (becoming more like closest_geq)

Since the we want to go through the range [from, to}, these additions were needed to guarantee that the start node is found.

gerard-ziemski reviewed (commented):
Nitpick: everywhere else in our code we refer to the tag parameter as `mem_tag`, not `mt`. I'm not 100% sure whether this needs to be pointed out, but I think I'd prefer if we remained consistent and used `mem_tag`.

gerard-ziemski commented:
I remember Lois asked, suring your presentation at the staff meeting, how to change our existing code to start using your RBTree and your answer was a single line change. Could you please remind me what that line was? I'd like to benchmark the new RBTree using my NMT benchmark to see whether there is any performance difference from Treap.

caspernorrbin commented:
> I remember Lois asked, during your presentation at the staff meeting, how to change our existing code to start using your RBTree and your answer was a single line change. Could you please remind me what that line was? I'd like to benchmark the new RBTree using my NMT benchmark to see whether there is any performance difference from Treap.


To swap out to the red-black tree, you can change these two lines in `vmatree.hpp`:
```c++
  using VMATreap = TreapCHeap<position, IntervalChange, PositionComparator>;
  using TreapNode = VMATreap::TreapNode;
```
to:
```c++
  using VMATreap = RBTreeCHeap<position, IntervalChange, PositionComparator, mtNMT>;
  using TreapNode = VMATreap::RBNode;
```

The internal variable names would still refer to the treap, but for testing that shouldn't matter.
If you want the VMATree gtests to compile, you need to do two more things:
Add `friend class NMTVMATreeTest;` to the `RBTree` class.
Change `treap.find(treap._root, key);` to `treap.find_node(key);` in `test_vmatree.cpp:54`

I have also done some benchmarking on my own only comparing the two trees in isolation. In the worst case scenario, which is only sequential inserts/removals, the red-black tree was a bit slower (~10%), but with more balanced operations, the treap was slower. Hopefully you will find similar results. I don't know how large of a % of NMT overhead is tree operations. If it's not high it shouldn't affect overall performance much in either direction.

gerard-ziemski commented:
> > I remember Lois asked, during your presentation at the staff meeting, how to change our existing code to start using your RBTree and your answer was a single line change. Could you please remind me what that line was? I'd like to benchmark the new RBTree using my NMT benchmark to see whether there is any performance difference from Treap.
> 
> To swap out to the red-black tree, you can change these two lines in `vmatree.hpp`:
> 
> ```c++
>   using VMATreap = TreapCHeap<position, IntervalChange, PositionComparator>;
>   using TreapNode = VMATreap::TreapNode;
> ```
> 
> to:
> 
> ```c++
>   using VMATreap = RBTreeCHeap<position, IntervalChange, PositionComparator, mtNMT>;
>   using TreapNode = VMATreap::RBNode;
> ```
> 
> The internal variable names would still refer to the treap, but for testing that shouldn't matter. If you want the VMATree gtests to compile, you need to do two more things: Add `friend class NMTVMATreeTest;` to the `RBTree` class. Change `treap.find(treap._root, key);` to `treap.find_node(key);` in `test_vmatree.cpp:54`
> 
> I have also done some benchmarking on my own only comparing the two trees in isolation. In the worst case scenario, which is only sequential inserts/removals, the red-black tree was a bit slower (~10%), but with more balanced operations, the treap was slower. Hopefully you will find similar results. I don't know how large of a % of NMT overhead is tree operations. If it's not high it shouldn't affect overall performance much in either direction.

Thank you! I don't need gtest, so the 2 line change will work for me. I will report back my findings after I run my benchmark...

gerard-ziemski commented:
And this:

```
* For target hotspot_variant-server_libjvm_objs_memoryFileTracker.o:
In file included from /Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/memoryFileTracker.cpp:1:
In file included from /Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/precompiled/precompiled.hpp:45:
In file included from /Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/memTracker.hpp:30:
In file included from /Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/memoryFileTracker.hpp:32:
/Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/vmatree.hpp:235:11: error: no matching member function for call to 'visit_in_order'
  235 |     _tree.visit_in_order(f);
      |     ~~~~~~^~~~~~~~~~~~~~
/Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/memoryFileTracker.cpp:75:15: note: in instantiation of function template specialization 'VMATree::visit_in_order<(lambda at /Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/memoryFileTracker.cpp:75:30)>' requested here
   75 |   file->_tree.visit_in_order([&](VMATree::TreapNode* current) {
      |               ^
/Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/utilities/rbTree.hpp:249:8: note: candidate function template not viable: 'this' argument has type 'const VMATreap' (aka 'const RBTree<unsigned long, VMATree::IntervalChange, VMATree::PositionComparator, RBTreeCHeapAllocator<(MemTag)'\f'>>'), but method is not marked const
  249 |   void visit_in_order(F f);
      |        ^
1 error generated.

```

caspernorrbin commented:
@gerard-ziemski I ran in to both of these trying to verify that swapping still worked. 
I pushed an update earlier today to fix the second issue, the `const` of `visit_in_order` had disappeared when I moved functions around. For the first issue, you would need to `#include rbTree.inline.hpp` with/instead of `rbTree.hpp`  in vmatree.

gerard-ziemski commented:
Hmm, I still can't compile it, I see:

```
/Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/vmatree.cpp:235:18: error: expected ';' after expression
  235 |   VMATreap::Range range(nullptr, nullptr);
      |                  ^
      |                  ;
/Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/vmatree.cpp:235:13: error: no member named 'Range' in 'RBTree<unsigned long, VMATree::IntervalChange, VMATree::PositionComparator, RBTreeCHeapAllocator<MemTag::mtNMT>>'
  235 |   VMATreap::Range range(nullptr, nullptr);
      |   ~~~~~~~~~~^
/Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/vmatree.cpp:235:19: error: use of undeclared identifier 'range'
  235 |   VMATreap::Range range(nullptr, nullptr);
      |                   ^
/Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/vmatree.cpp:240:5: error: use of undeclared identifier 'range'
  240 |     range = _tree.find_enclosing_range(from);
      |     ^
/Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/vmatree.cpp:240:19: error: no member named 'find_enclosing_range' in 'RBTree<unsigned long, VMATree::IntervalChange, VMATree::PositionComparator, RBTreeCHeapAllocator<MemTag::mtNMT>>'
  240 |     range = _tree.find_enclosing_range(from);
      |             ~~~~~ ^
/Users/gerard/Work/bugs/8317453/jdk/src/hotspot/share/nmt/vmatree.cpp:241:10: error: use of undeclared identifier 'range'
  241 |     if ((range.start == nullptr && range.end == nullptr) ||
      |          ^

```

caspernorrbin commented:
@gerard-ziemski That one is on me! I still had an old copy of VMATree that didn't have `find_enclosing range`. I've added it now and tested with the latest version, so hopefully there should not be any more issues.

gerard-ziemski commented:
From the point of view of my benchmark, using data recorded from a java server with 10k accesses, the performance looks identical. I ran benchmark 100 times for each and chose the fastest run:

`time:194,279[ns] [samples:1,803]` vs `time:194,516[ns] [samples:1,803]`

No recursion and deterministic behavior (no dependency on 64bit random), however, in my opinion gives RBTree an advantage over Treaps. Good job!

caspernorrbin commented:
@tstuefe Did you have any more feedback on this? Would like to get this pushed soon so I can publish the intrusive variant.

tstuefe commented:
> @tstuefe Did you have any more feedback on this? Would like to get this pushed soon so I can publish the intrusive variant.

Sorry, I have not forgotten you, but am booked out with FOSDEM preparations. I can continue the review when I come back (end of next week), or the week after.

I would have liked to prod at the meat of the insertion/deletion logic more, since I remember that part being particularly fiddly. We should get this right, this needs close review.

But I don't want to hold you up, so if you feel you need to push now, go ahead. I would definitely like to look at the intrusive variant, since I expect I will use that one quite a bit.

tstuefe reviewed (approved):
Approved, good work

caspernorrbin commented:
> 
> Sorry, I have not forgotten you, but am booked out with FOSDEM preparations. I can continue the review when I come back (end of next week), or the week after.
> 
> I would have liked to prod at the meat of the insertion/deletion logic more, since I remember that part being particularly fiddly. We should get this right, this needs close review.
> 
> But I don't want to hold you up, so if you feel you need to push now, go ahead. I would definitely like to look at the intrusive variant, since I expect I will use that one quite a bit.

Thank you! The intrusive tree is [done](/caspernorrbin/jdk/tree/rb-tree-intrusive), waiting for this PR. It keeps the insertion/deletion logic identical, but reworks a lot of the surrounding functions. Since the tree isn't used anywhere yet, perhaps you could have a closer look at the insertion/deletion logic then? If there are any issues we can fix it there before the tree gets sees any real use.

xmas92 commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -158,95 +168,126 @@ class RBTree {
   }
 
+  const RBNode* closest_leq(const K& k, BoundMode mode = INCLUSIVE) const {
+    return const_cast<RBTree<K, V, COMPARATOR, ALLOCATOR>*>(this)->closest_leq(k, mode);
```
When it comes to `const` overloads of `accessor` like member functions I prefer that we invert the `const` cast logic. That is that it is the `const` variant that includes the implementation. And that the non-`const` variant dispatches to the `const` versions and then casts away the the `const` qualifier on the result. This ensures that the implementation does not break the `const` property of the class, and is less bug prone.

caspernorrbin replied:
Swapped them around to have the non-const depend on the const version

xmas92 commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -158,95 +168,126 @@ class RBTree {
   // Finds the node with the closest key <= the given key
-  RBNode* closest_leq(const K& key) {
+  // Change mode to EXCLUSIVE to not include node matching key
+  RBNode* closest_leq(const K& key, BoundMode mode = INCLUSIVE) {
```
Having a parameter value change the behaviour to contradict the name of the function feels a little evil. Might cause confusion and bugs in the future. I would rather have one `closest(Key, Direction, BoundMode)` and/or four `closest_[ls,leq,geq,gt](Key)`.

xmas92 commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -158,95 +168,126 @@ class RBTree {
+  }
+
+  const RBNode* closest_gt(const K& k, BoundMode mode = EXCLUSIVE) const {
+    return const_cast<RBTree<K, V, COMPARATOR, ALLOCATOR>*>(this)->closest_gt(k, mode);
```
Same.

xmas92 commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -158,95 +168,126 @@ class RBTree {
+  RBNode* find_node(const K& k);
+
+  const RBNode* find_node(const K& k) const {
+    return const_cast<RBTree<K, V, COMPARATOR, ALLOCATOR>*>(this)->find_node(k);
```
Same.

xmas92 commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -55,61 +54,72 @@ class RBTree {
-        : _parent(nullptr), _left(nullptr), _right(nullptr),
-          _key(k), _value(v), _color(RED) {}
+    RBNode* parent() const { return (RBNode*)(_parent & ~0x1); }
+    void set_parent(RBNode* new_parent) {_parent = (_parent & 0x1) | ((uintptr_t)new_parent & ~0x1); }
```
When would `new_parent` not have a valid `RBNode*` pointer value? AFAICT it is only called with valid pointers. The lsb masking is unnecessary. If it is actually required from somewhere, I missed it, in which case we should use `uintptr_t`.

jdksjolen replied:
You just missed it, if I understand you correctly. To quote the code: `uintptr_t _parent; // LSB encodes color information. 0 = RED, 1 = BLACK`.

caspernorrbin replied:
You're right, the masking here does nothing. I removed it.

xmas92 reviewed (changes requested):
This looks really good now. Just a few more comments.

The parameter names should also be harmonised. At least for the public member functions. Right now both `k` and `key` is used, we should pick one (I like descriptive names in signatures, so I would pick the later, but I think the former use is more prevalent in your patch). And it should be consistent with `v`/`value`.

xmas92 commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -0,0 +1,541 @@
+#include "utilities/debug.hpp"
+#include "utilities/powerOfTwo.hpp"
+#include "utilities/rbTree.hpp"
+#include <utility> // for std::swap
```
Can we use the `swap` from `utilities/globalDefinitions.hpp` instead? This would be the first explicit include of `<utility>` in hotspot. It is probably the path of least resistance.

xmas92 reviewed (approved):
lgtm.

caspernorrbin commented:
Thank you everyone for taking the time to review.

/integrate

tstuefe commented:
/sponsor
